runelite-client/src/main/java/net/runelite/client/chat/ChatCommandManager.java

```diff
@@ -0,0 +1,198 @@
+
+	@Inject
+	public ChatCommandManager(EventBus eventBus, CommandManager commandManager)
+	{
+		eventBus.register(this);
+		commandManager.register(this);
+	}
+
+	public void registerCommand(String command, BiConsumer<SetMessage, String> execute)
+	{
+		registerCommand(command, execute, null);
+	}
+
+	public void registerCommand(String command, BiConsumer<SetMessage, String> execute, BiPredicate<ChatInput, String> input)
+	{
+		commands.put(command.toLowerCase(), new ChatCommand(command, false, execute, input));
+	}
+
+	public void registerCommandAsync(String command, BiConsumer<SetMessage, String> execute)
+	{
+		registerCommandAsync(command, execute, null);
+	}
+
+	public void registerCommandAsync(String command, BiConsumer<SetMessage, String> execute, BiPredicate<ChatInput, String> input)
+	{
+		commands.put(command.toLowerCase(), new ChatCommand(command, true, execute, input));
+	}
+
+	public void unregisterCommand(String command)
+	{
+		commands.remove(command.toLowerCase());
+	}
+
+	@Subscribe
+	public void onSetMessage(SetMessage setMessage)
+	{
+		if (client.getGameState() != GameState.LOGGED_IN)
+		{
+			return;
+		}
+
+		switch (setMessage.getType())
+		{
+			case PUBLIC:
+			case PUBLIC_MOD:
+			case CLANCHAT:
+			case PRIVATE_MESSAGE_RECEIVED:
+			case PRIVATE_MESSAGE_SENT:
+				break;
+			default:
+				return;
+		}
+
+		String message = setMessage.getValue();
+
+		String command = extractCommand(message);
+		if (command == null)
+		{
+			return;
+		}
+
+		ChatCommand chatCommand = commands.get(command.toLowerCase());
+		if (chatCommand == null)
+		{
+			return;
+		}
+
+		if (chatCommand.isAsync())
+		{
+			scheduledExecutorService.execute(() -> chatCommand.getExecute().accept(setMessage, message));
+		}
+		else
+		{
+			chatCommand.getExecute().accept(setMessage, message);
+		}
+	}
+
+	@Override
+	public boolean onChatboxInput(ChatboxInput chatboxInput)
+	{
+		String message = chatboxInput.getValue();
+		if (message.startsWith("/"))
+		{
+			message = message.substring(1); // clan chat input
+		}
+
+		String command = extractCommand(message);
+		if (command == null)
+		{
+			return false;
+		}
+
+		ChatCommand chatCommand = commands.get(command.toLowerCase());
+		if (chatCommand == null)
+		{
+			return false;
+		}
+
+		BiPredicate<ChatInput, String> input = chatCommand.getInput();
+		if (input == null)
+		{
+			return false;
+		}
+
+		return input.test(chatboxInput, message);
+	}
+
+	@Override
+	public boolean onPrivateMessageInput(PrivateMessageInput privateMessageInput)
+	{
+		final String message = privateMessageInput.getMessage();
+
+		String command = extractCommand(message);
+		if (command == null)
+		{
+			return false;
+		}
+
+		ChatCommand chatCommand = commands.get(command.toLowerCase());
+		if (chatCommand == null)
+		{
+			return false;
+		}
+
+		BiPredicate<ChatInput, String> input = chatCommand.getInput();
+		if (input == null)
+		{
+			return false;
+		}
+
+		return input.test(privateMessageInput, message);
```

Comment:
I would also substr command out of message, but based on Discord discussion !total and !lvl needs it, because they are using same callback, but that can be easily solved by making 2 methods that call 1 shared one (or something else)